File: services/forecast-engine/src/forecast_engine/adapters/object_store.py
"""
Adaptador MinIO/S3 do motor de previsão (D5, D18).

URIs seguem o formato s3://bucket/key/path conforme contracts/messaging.md.
O motor lê datasets em Parquet, grava resultados em Parquet e escreve o
marcador _SUCCESS após concluir — garantindo que a idempotência (D6) não
republique um resultado escrito pela metade.
"""

from __future__ import annotations

from io import BytesIO
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from mypy_boto3_s3 import S3Client
else:
    S3Client = object


def _parse_uri(uri: str) -> tuple[str, str]:
    """Parse s3://bucket/key para (bucket, key)."""
    if not uri.startswith("s3://"):
        raise ValueError(f"URI inválida (esperado s3://): {uri!r}")
    without_scheme = uri[5:]
    slash = without_scheme.index("/")
    bucket = without_scheme[:slash]
    key = without_scheme[slash + 1 :]
    return bucket, key


class ObjectStore:
    """Acesso ao MinIO/S3 para o motor de previsão."""

    def __init__(self, client: S3Client) -> None:
        self._client = client

    # ─── Leitura ──────────────────────────────────────────────────

    def get_bytes(self, uri: str) -> bytes:
        """Baixa o objeto completo como bytes."""
        bucket, key = _parse_uri(uri)
        response = self._client.get_object(Bucket=bucket, Key=key)
        return response["Body"].read()

    # ─── Escrita ──────────────────────────────────────────────────

    def put_bytes(self, uri: str, data: bytes) -> None:
        """Envia bytes para o MinIO."""
        bucket, key = _parse_uri(uri)
        self._client.put_object(Bucket=bucket, Key=key, Body=data)

    def write_success_marker(self, output_prefix: str) -> None:
        """
        Grava o marcador _SUCCESS após escrever os Parquets (D18).

        Deve ser chamado por último, só após output.parquet e series.parquet
        estarem persistidos. A idempotência verifica este objeto antes de
        recalcular (D6) — se ele existe, republica o resultado em vez de
        recalcular.
        """
        sep = "" if output_prefix.endswith("/") else "/"
        self.put_bytes(f"{output_prefix}{sep}_SUCCESS", b"")

    # ─── Verificação de conclusão ──────────────────────────────────

    def is_complete(self, output_prefix: str) -> bool:
        """Verifica se o marcador _SUCCESS existe (idempotência D6)."""
        sep = "" if output_prefix.endswith("/") else "/"
        bucket, key = _parse_uri(f"{output_prefix}{sep}_SUCCESS")
        try:
            self._client.head_object(Bucket=bucket, Key=key)
            return True
        except self._client.exceptions.ClientError as exc:
            code = exc.response["Error"]["Code"]  # type: ignore[attr-defined]
            if code in ("404", "NoSuchKey"):
                return False
            raise

File: apps/api/src/adapters/minio/object-store.ts
import { GetObjectCommand, PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import type { DatasetStore } from '../../composition/ports.js';

/**
 * Adaptador MinIO/S3 (D5, D18).
 *
 * URIs seguem o formato `s3://bucket/key/path` conforme contracts/messaging.md.
 * Upload em streaming — nunca bufferiza o dataset na memória da API (D7).
 */

function parseUri(uri: string): { bucket: string; key: string } {
  // s3://bucket/key/path → { bucket, key }
  const withoutScheme = uri.slice(5); // remove "s3://"
  const slash = withoutScheme.indexOf('/');
  if (slash === -1) throw new Error(`URI inválida: ${uri}`);
  return {
    bucket: withoutScheme.slice(0, slash),
    key: withoutScheme.slice(slash + 1),
  };
}

export class MinioObjectStore implements DatasetStore {
  constructor(private readonly client: S3Client) {}

  async putStream(uri: string, stream: NodeJS.ReadableStream): Promise<void> {
    const { bucket, key } = parseUri(uri);
    await this.client.send(
      new PutObjectCommand({ Bucket: bucket, Key: key, Body: stream as never }),
    );
  }

  async presignGet(uri: string, expiresInSeconds: number): Promise<string> {
    const { bucket, key } = parseUri(uri);
    const command = new GetObjectCommand({ Bucket: bucket, Key: key });
    return getSignedUrl(this.client, command, { expiresIn: expiresInSeconds });
  }

  /**
   * Grava marcador `_SUCCESS` após os Parquets (D18).
   * Idempotência do motor verifica este objeto antes de recalcular (D6).
   */
  async writeSuccessMarker(outputPrefix: string): Promise<void> {
    const uri = outputPrefix.endsWith('/') ? `${outputPrefix}_SUCCESS` : `${outputPrefix}/_SUCCESS`;
    const { bucket, key } = parseUri(uri);
    await this.client.send(new PutObjectCommand({ Bucket: bucket, Key: key, Body: '' }));
  }
}

export function createS3Client(): S3Client {
  const endpoint = process.env.S3_ENDPOINT;
  return new S3Client({
    ...(endpoint !== undefined ? { endpoint } : {}),
    region: process.env.S3_REGION ?? 'us-east-1',
    credentials: {
      accessKeyId: process.env.S3_ACCESS_KEY ?? '',
      secretAccessKey: process.env.S3_SECRET_KEY ?? '',
    },
    forcePathStyle: true, // obrigatório para MinIO
  });
}
